### ISSUE TITLE
Godot 4 FPS Support

### ISSUE THREAD

**Issue by cadennabors** [lawnjelly/smoothing-addon#51]
When using the 3d smoothing node on the head of my character, it works really well to smooth the movement instead of jittering at physics speed. The problem is :

Camera rotation is not updated in physics ticks!

When using a smoothing node before the camera, then rotations are smoothed rather than keeping their updated values between physics frames.

When disabling 'basis' on the smoothing node, rotation will not update with the parents, removing rotation altogether.

This makes setting up an FPS controller that doesn't have jitter very complicated.

**Comment by cadennabors** [comment]
A simple fix to this issue would be adding a new parameter called something like "keep previous transform". This allows rotation and scale of the node to be edited, while still not affecting any things that are enabled.

Under process, it would be 

```
func _process(_delta):

	var f = Engine.get_physics_interpolation_fraction()
	var tr: Transform3D
	if SF_KEEPTRANSFORM:
		tr = transform
	else:
		tr = Transform3D()
```

instead of:

```
func _process(_delta):

	var f = Engine.get_physics_interpolation_fraction()
	var tr: Transform3D = Transform3D()
```

**Comment by lawnjelly** [comment]
I discuss this in this video for the core interpolation in 3.x:
https://www.youtube.com/watch?v=BJgDCISku_o&t=126s

For FPS camera and the addon, sometimes it can be better to keep the camera on a separate branch to the player, and manually interpolating in global space. Attaching the camera as a child of a (basis) smoothed player visual rep won't work well as you say, because the smoothed basis will likely be inherited from the parent, unless you turn off basis interpolation.

Typically in FPS you'd want to drive the rotation of the player (rather than the camera directly) using the mouse, so if you turn off basis interpolation in the player, and just alter the yaw of the player directly using the mouse (and thus inherited by a camera child) this may work for you. The pitch on the other hand might be applied directly to the camera, so the player rep is always level but the camera looks up and down. The exact setup depends on the nature of your game / camera.

**Comment by cadennabors** [comment]
That is correct, but it does allow for some more freedom. With this approach, my "Player" node can stay as a CharacterBody3D, which allows positional
smoothing without smoothing its rotation, allowing for exact control over the camera.

**Comment by lawnjelly** [comment]
Ah sorry, I get what you mean. This may well be an oversight on my part, will look into it when I get a moment. :+1: :smiley:

**Comment by bryanmylee** [comment]
I've managed to solve this problem by modifying the addon. Since my player is translated on the physics tick but rotated on the process tick, I needed a way to selectively apply interpolation on a per-property basis. To do so, I made this change:

```gdscript
func _process(_delta):

	var f = Engine.get_physics_interpolation_fraction()
	var tr: Transform3D = Transform3D()

	# translate
	if _TestFlags(SF_TRANSLATE):
		var ptDiff = _m_trCurr.origin - _m_trPrev.origin
		tr.origin = _m_trPrev.origin + (ptDiff * f)
+	else:
+		tr.origin = _m_Target.global_transform.origin

	# rotate
	if _TestFlags(SF_BASIS):
		if _TestFlags(SF_SLERP):
			tr.basis = _m_trPrev.basis.slerp(_m_trCurr.basis, f)
		else:
			tr.basis = _LerpBasis(_m_trPrev.basis, _m_trCurr.basis, f)
+	else:
+		tr.basis = _m_Target.global_transform.basis

	transform = tr
```

I'm not sure if this was the intention initially, but disabling the `basis` or `translate` flag OOTB causes the transform to never be applied at all. With the change, I can configure the smoothing node to only interpolate `translate` but leave `basis` as-is.
